osquery/tables/system/linux/pci_devices.cpp

```diff
@@ -26,38 +31,191 @@ const std::string kPCIKeyVendor = "ID_VENDOR_FROM_DATABASE";
+  return vendor(vendor_id, [&](const PciVendor& v) {
+    name = v.name;
+    return Status::success();
+  });
+}
+
+Status PciDB::getModel(const std::string& vendor_id,
+                       const std::string& model_id,
+                       std::string& model) {
+  return vendor(vendor_id, [&](const PciVendor& v) {
+    auto model_it = v.models.find(model_id);
+    if (model_it == v.models.end()) {
+      return Status::failure("Model ID does not exist");
+    }
+
+    model = model_it->second.desc;
+
+    return Status::success();
+  });
+}
+
+Status PciDB::getSubsystemInfo(const std::string& vendor_id,
+                               const std::string& model_id,
+                               const std::string& subsystem_vendor_id,
+                               const std::string& subsystem_device_id,
+                               std::string& subsystem) {
+  return vendor(vendor_id, [&](const PciVendor& v) {
+    auto model_it = v.models.find(model_id);
+    if (model_it == v.models.end()) {
+      return Status::failure("Model ID does not exist");
+    }
+
+    auto subsystem_id = subsystem_vendor_id + " " + subsystem_device_id;
+
+    auto subsystem_it = model_it->second.subsystemInfo.find(subsystem_id);
+    if (subsystem_it == model_it->second.subsystemInfo.end()) {
+      return Status::failure("Subsystem ID does not exist in system pci.ids: " +
+                             subsystem_id);
+    }
+
+    subsystem = subsystem_it->second;
+
+    return Status::success();
+  });
+}
+
+Status PciDB::vendor(const std::string& vendor_id,
+                     std::function<Status(const PciVendor&)> predicate) {
+  auto vendor_it = db_.find(vendor_id);
+  if (vendor_it == db_.end()) {
+    return Status::failure("Vendor ID does not exist");
+  }
+
+  return predicate(vendor_it->second);
+}
 
 QueryData genPCIDevices(QueryContext& context) {
   QueryData results;
 
-  auto delUdev = [](udev* u) { udev_unref(u); };
-  std::unique_ptr<udev, decltype(delUdev)> udev_handle(udev_new(), delUdev);
+  auto del_udev = [](udev* u) { udev_unref(u); };
+  std::unique_ptr<udev, decltype(del_udev)> udev_handle(udev_new(), del_udev);
   if (udev_handle.get() == nullptr) {
     VLOG(1) << "Could not get udev handle";
     return results;
   }
 
   // Perform enumeration/search.
-  auto delUdevEnum = [](udev_enumerate* e) { udev_enumerate_unref(e); };
-  std::unique_ptr<udev_enumerate, decltype(delUdevEnum)> enumerate(
-      udev_enumerate_new(udev_handle.get()), delUdevEnum);
+  auto del_udev_enum = [](udev_enumerate* e) { udev_enumerate_unref(e); };
+  std::unique_ptr<udev_enumerate, decltype(del_udev_enum)> enumerate(
+      udev_enumerate_new(udev_handle.get()), del_udev_enum);
   if (enumerate.get() == nullptr) {
     VLOG(1) << "Could not get udev_enumerate handle";
     return results;
   }
 
+  std::ifstream raw(kPciIdsPath);
+  if (raw.fail()) {
+    LOG(ERROR) << "failed to read " << kPciIdsPath;
```

Comment:
Could you add more information?  I like your format of the error message in other places mentioning pci.ids